uhubctl.c

```diff
@@ -1051,13 +1092,427 @@ static int usb_find_hubs(void)
+        cJSON_AddStringToObject(port_json, "usb_version", usb_version);
+
+        // Add device version
+        char device_version[8];
+        snprintf(device_version, sizeof(device_version), "%x.%02x", desc.bcdDevice >> 8, desc.bcdDevice & 0xFF);
+        cJSON_AddStringToObject(port_json, "device_version", device_version);
+
+        // Add number of configurations
+        cJSON_AddNumberToObject(port_json, "num_configurations", desc.bNumConfigurations);
+
+        // Check if it's a mass storage device
+        cJSON_AddBoolToObject(port_json, "is_mass_storage", is_mass_storage_device(dev));
+    }
+
+    // Add description from the descriptor strings
+    cJSON_AddStringToObject(port_json, "description", ds->description);
+}
+
+
+// Main function to create port status JSON
+cJSON* create_port_status_json(int port, int port_status, const struct descriptor_strings* ds, struct libusb_device *dev)
+{
+    cJSON* port_json = cJSON_CreateObject();
+    cJSON_AddNumberToObject(port_json, "port", port);
+    
+    char status_hex[7];
+    snprintf(status_hex, sizeof(status_hex), "0x%04x", port_status);
+    cJSON_AddStringToObject(port_json, "status", status_hex);
+    
+    cJSON_AddItemToObject(port_json, "flags", create_status_flags_json(port_status));
+    cJSON_AddItemToObject(port_json, "human_readable", create_human_readable_json(port_status));
+
+    char* speed_str;
+    int64_t speed_bits;
+    get_port_speed(port_status, &speed_str, &speed_bits);
+    cJSON_AddStringToObject(port_json, "speed", speed_str);
+    cJSON_AddNumberToObject(port_json, "speed_bits", (double)speed_bits);
+   
+    if (port_status & USB_PORT_STAT_CONNECTION) {
+        add_device_info_json(port_json, dev, ds);
+    }
+
+    return port_json;
+}
+cJSON* create_hub_json(struct hub_info* hub, int portmask)
+{
+    cJSON* hub_json = cJSON_CreateObject();
+    cJSON_AddStringToObject(hub_json, "location", hub->location);
+    cJSON_AddStringToObject(hub_json, "description", hub->ds.description);
+    
+    cJSON* hub_info = cJSON_CreateObject();
+    
+    unsigned int vendor_id, product_id;
+    sscanf(hub->vendor, "%x:%x", &vendor_id, &product_id);
+    
+    char vendor_id_hex[8], product_id_hex[8];
+    snprintf(vendor_id_hex, sizeof(vendor_id_hex), "0x%04x", vendor_id);
+    snprintf(product_id_hex, sizeof(product_id_hex), "0x%04x", product_id);
+    
+    cJSON_AddStringToObject(hub_info, "vendor_id", vendor_id_hex);
+    cJSON_AddStringToObject(hub_info, "product_id", product_id_hex);
+    
+    char usb_version[16];
+    snprintf(usb_version, sizeof(usb_version), "%x.%02x", hub->bcd_usb >> 8, hub->bcd_usb & 0xFF);
+    cJSON_AddStringToObject(hub_info, "usb_version", usb_version);
+    
+    cJSON_AddNumberToObject(hub_info, "num_ports", hub->nports);
+    
+    const char* power_switching_mode;
+    switch (hub->lpsm) {
+        case HUB_CHAR_INDV_PORT_LPSM:
+            power_switching_mode = "per-port";
```

Comment:
This mode is already set to `ppps` in get_device_description, make it consistent?